nomad/acl_endpoint.go

```diff
@@ -1264,13 +1264,38 @@ func (a *ACL) ListRoles(
 	}
 	defer metrics.MeasureSince([]string{"nomad", "acl", "list_roles"}, time.Now())
 
-	// TODO (jrasell) allow callers to list role associated to their token.
-	if acl, err := a.srv.ResolveToken(args.AuthToken); err != nil {
+	// Resolve the token and ensure it has some form of permissions.
+	acl, err := a.srv.ResolveToken(args.AuthToken)
+	if err != nil {
 		return err
-	} else if acl == nil || !acl.IsManagement() {
+	} else if acl == nil {
 		return structs.ErrPermissionDenied
 	}
 
+	// If the token is a management token, they can list all tokens. If not,
+	// the roles mapping tracks which role links the token has and therefore
+	// which ones the caller can list.
+	isManagement := acl.IsManagement()
+	var roles map[string]struct{}
+
+	// If the token is not a management token, we determine which roles are
+	// linked to the token and therefore can be listed by the caller.
+	if !isManagement {
+		token, err := a.requestACLToken(args.AuthToken)
+		if err != nil {
+			return err
+		}
+		if token == nil {
+			return structs.ErrTokenNotFound
+		}
+
+		roles = make(map[string]struct{}, len(token.Roles))
+
+		for _, roleLink := range token.Roles {
+			roles[roleLink.ID] = struct{}{}
+		}
```

Comment:
just tagged v0.1.3 of go-set to include FromFunc, to turn all this into one line